src/mpid/ch4/src/ch4_coll.h

```diff
@@ -408,6 +408,13 @@ MPL_STATIC_INLINE_PREFIX int MPID_Allreduce(const void *sendbuf, void *recvbuf,
 
     is_commutative = MPIR_Op_is_commutative(op);
 
+    MPI_Aint dt_size;
+    MPIR_Datatype_get_size_macro(datatype, dt_size);
+    if (dt_size >=
+        MPIR_CVAR_REDUCE_INTRANODE_BUFFER_TOTAL_SIZE / MPIR_CVAR_REDUCE_INTRANODE_NUM_CELLS) {
+        goto fallback;
```

Comment:
Not sure if this is correct,  Not all composition algorithms do pipelining I think. Maybe this check should go one level down to the alpha, beta gamma functions?

Reply:
That might be an improvement, but will take some time to investigate. I'm tempted to pick this patch as-is for 4.0.x and ship since it fixes a user-reported bug. I guess we have a little extra time to get it right for 4.1.

@hzhou the commit message mentions MPID_Bcast, but this patch only modified reduce and allreduce. Is there a missing piece?

Reply:
I originally thought bcast was susceptible, but it turns out bcast can simply break down to basic elements, so the outmost type size is irrelevant. Only reduce and allreduce are affected. Let me update the commit message...